README.md
# apt-transport-s3

Additional "s3" protocol for apt so you can host your giant apt repository in s3 on the cheap!

We use this for pressflip.com to deploy and distribute all of our software.  apt is a great packaging system and s3 is a great place to backup/store static files.  apt-s3 is especially useful and fast if you are hosting your servers within EC2.

Current Author: Zachary Schneider @sigil66

Original Author: Kyle Shank

Contributors: Cliff Moon (@cliffmoon on GH), Jens Braeuer (@jbraeuer)

Documenter: Susan Potter (@mbbx6spp on GH)

## Building

Before building this project on Ubuntu (tested on 12.04) you will need to install the following packages:

    [sudo] apt-get install libapt-pkg-dev libcurl4-openssl-dev

To build this project you simply run `make`. It will produce a binary named `s3` under the `src/` dir.

## Installing

Once compiled, the resulting s3 binary must be placed in /usr/lib/apt/methods/ along with the other protocol binaries.

Finally, this is how you add it to the /etc/apt/sources.list file if you want your credentials in the url:

    deb s3://AWS_ACCESS_ID:[AWS_SECRET_KEY_IN_BRACKETS]@s3-ENDPOINT.amazonaws.com/BUCKETNAME prod main
    
Https:
    
    deb s3+https://AWS_ACCESS_ID:[AWS_SECRET_KEY_IN_BRACKETS]@s3-ENDPOINT.amazonaws.com/BUCKETNAME prod main
    

otherwise leave off the credentials and it will draw them from the environment variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_KEY_ID`.

Simply upload all of your .deb packages and Packages.gz file into the s3 bucket you chose with the file key mapping that matches the file system layout.

## Uploading repository to s3

Before synchronization, you need a s3cmd tool installed and configured:

    [sudo] apt-get install s3cmd
    
    s3cmd --configure

To synchronize local repository to s3 as read-only, execute:

    s3cmd sync /srv/apt-repo-dir/dists s3://bucket_name
    s3cmd sync /srv/apt-repo-dir/pool s3://bucket_name
    
## Using GPG keys

If you're signing you repository with key, export it to server:

    gpg --send-keys XXXXXXXX

then import it and install to apt:

    gpg --recv-keys XXXXXXXX
    gpg -a --export XXXXXXXX | [sudo] apt-key add -


debian/changelog
apt-transport-s3 (1.1.3) precise; urgency=low
  
  * Change repository name
  * Add auth support to https method
  
 -- Zachary Schneider <sigil.66@gmail.com>  Tues, 09 Sept 2014 15:43:00 -0500

apt-transport-s3 (1.1.2) precise; urgency=low
 
  * Break out http handling
  * Add https handler

 -- Zachary Schneider <sigil.66@gmail.com>  Tues, 09 Oct 2012 21:00:00 -0500

apt-transport-s3 (1.1.1ubuntu2) precise; urgency=low

  * Modified for company needs

 -- Piotr Bulinski <piotr.bulinski@castlabs.com>  Fri, 27 Jun 2012 17:00:03 +0200

apt-transport-s3 (1.1.1ubuntu1) precise; urgency=low

  * Version bump for stvp; no change from braeuer.jens,
    use his package instead.

 -- Brian P O'Rourke <bpo@somnambulance.net>  Fri, 04 May 2012 16:56:03 -0700

apt-transport-s3 (1.1ubuntu1) oneiric; urgency=low

  * Make apt-transport-s3 work, when locale is non-english.

 -- Jens Braeuer <braeuer.jens@googlemail.com>  Mon, 31 Jan 2012 21:01:25 +0100

apt-transport-s3 (1.0ubuntu1) lucid; urgency=low

  * Working around stupid amazon s3 behaviour with filenames
    that include + signs

 -- Jorge Gallegos <kad@blegh.net>  Wed, 18 May 2011 14:01:25 -0700

apt-transport-s3 (1.0) unstable; urgency=low

  * Initial release.

 -- Jorge Gallegos <kad@blegh.net>  Fri, 29 Apr 2011 15:20:02 -0700


src/s3+https.h
// -*- mode: cpp; mode: fold -*-
// Description								/*{{{*/// $Id: http.h,v 1.12 2002/04/18 05:09:38 jgg Exp $
// $Id: http.h,v 1.12 2002/04/18 05:09:38 jgg Exp $
/* ######################################################################

   HTTP Acquire Method - This is the HTTP aquire method for APT.

   ##################################################################### */
                  /*}}}*/

#ifndef APT_HTTPS_H
#define APT_HTTPS_H

#include <iostream>
#include <curl/curl.h>

using std::cout;
using std::endl;

class HttpsMethod;
class FileFd;

class HttpsMethod : public pkgAcqMethod
{
   // minimum speed in bytes/se that triggers download timeout handling
   static const int DL_MIN_SPEED = 10;

   virtual bool Fetch(FetchItem *);
   static size_t write_data(void *buffer, size_t size, size_t nmemb, void *userp);
   static int progress_callback(void *clientp, double dltotal, double dlnow, 
        double ultotal, double ulnow);
   void SetupProxy();
   CURL *curl;
   FetchResult Res;

   public:
   FileFd *File;

   HttpsMethod() : pkgAcqMethod("1.2",Pipeline | SendConfig) 
   {
      File = 0;
      curl = curl_easy_init();
   };
};

#include <apt-pkg/strutl.h>
URI Proxy;

#endif


src/s3+https.cc
//-*- mode: cpp; mode: fold -*-
// Description								/*{{{*/
// $Id: http.cc,v 1.59 2004/05/08 19:42:35 mdz Exp $
/* ######################################################################

   HTTPS Acquire Method - This is the HTTPS aquire method for APT.

   It uses libcurl

   ##################################################################### */
                  /*}}}*/
// Include Files							/*{{{*/
#include <config.h>

#include <apt-pkg/fileutl.h>
#include <apt-pkg/acquire-method.h>
#include <apt-pkg/error.h>
#include <apt-pkg/hashes.h>
#include <apt-pkg/netrc.h>
#include <apt-pkg/configuration.h>

#include <sys/stat.h>
#include <sys/time.h>
#include <utime.h>
#include <unistd.h>
#include <signal.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <iostream>
#include <sstream>

#include <openssl/hmac.h>
#include <openssl/evp.h>
#include <openssl/bio.h>

#include "config.h"
#include "s3+https.h"
#include <apti18n.h>

#define SLEN 1024

void doEncrypt(char * kString,char * str, const char * secretKey);
                  /*}}}*/
using namespace std;

size_t 
HttpsMethod::write_data(void *buffer, size_t size, size_t nmemb, void *userp)
{
   HttpsMethod *me = (HttpsMethod *)userp;

   if(me->File->Write(buffer, size*nmemb) != true)
      return false;

   return size*nmemb;
}

int 
HttpsMethod::progress_callback(void *clientp, double dltotal, double dlnow, 
            double ultotal, double ulnow)
{
   HttpsMethod *me = (HttpsMethod *)clientp;
   if(dltotal > 0 && me->Res.Size == 0) {
      me->Res.Size = (unsigned long long)dltotal;
      me->URIStart(me->Res);
   }
   return 0;
}

void HttpsMethod::SetupProxy()  					/*{{{*/
{
   URI ServerName = Queue->Uri;

   // Determine the proxy setting - try https first, fallback to http and use env at last
   string UseProxy = _config->Find("Acquire::https::Proxy::" + ServerName.Host,
           _config->Find("Acquire::http::Proxy::" + ServerName.Host).c_str());

   if (UseProxy.empty() == true)
      UseProxy = _config->Find("Acquire::https::Proxy", _config->Find("Acquire::http::Proxy").c_str());

   // User want to use NO proxy, so nothing to setup
   if (UseProxy == "DIRECT")
      return;

   if (UseProxy.empty() == false) 
   {
      // Parse no_proxy, a comma (,) separated list of domains we don't want to use
      // a proxy for so we stop right here if it is in the list
      if (getenv("no_proxy") != 0 && CheckDomainList(ServerName.Host,getenv("no_proxy")) == true)
   return;
   } else {
      const char* result = getenv("http_proxy");
      UseProxy = result == NULL ? "" : result;
   }

   // Determine what host and port to use based on the proxy settings
   if (UseProxy.empty() == false) 
   {
      Proxy = UseProxy;
      if (Proxy.Port != 1)
   curl_easy_setopt(curl, CURLOPT_PROXYPORT, Proxy.Port);
      curl_easy_setopt(curl, CURLOPT_PROXY, Proxy.Host.c_str());
   }
}									/*}}}*/

void doEncrypt(char *kString, char *sigString, const char* secretKey){
   HMAC_CTX hctx;
   BIO *bio, *b64;
   char *sigptr;
   long siglen = -1;
   unsigned int rizlen;
   char skey[SLEN];
   unsigned char results[SLEN];

   // Initialize SHA1 encryption
   sprintf(skey, "%s", secretKey);
   HMAC_CTX_init(&hctx);
   HMAC_Init(&hctx, skey, (int)strlen((char *)skey), EVP_sha1());

   // Encrypt
   HMAC(EVP_sha1(), skey, (int)strlen((char *)skey), (unsigned char *)kString,
         (int)strlen((char *)kString), results, &rizlen);

   // Base 64 Encode
   b64 = BIO_new(BIO_f_base64());
   bio = BIO_new(BIO_s_mem());
   BIO_set_flags(b64, BIO_FLAGS_BASE64_NO_NL);
   bio = BIO_push(b64, bio);
   BIO_write(bio, results, rizlen);
   BIO_flush(bio);

   siglen = BIO_get_mem_data(bio, &sigptr);
   memcpy(sigString, sigptr, siglen);
   sigString[siglen] = '\0';

   // Clean up Encryption, Encoding
   BIO_free_all(bio);
   HMAC_CTX_cleanup(&hctx);
}

// HttpsMethod::Fetch - Fetch an item					/*{{{*/
// ---------------------------------------------------------------------
/* This adds an item to the pipeline. We keep the pipeline at a fixed
   depth. */
bool HttpsMethod::Fetch(FetchItem *Itm)
{
   struct stat SBuf;
   struct curl_slist *headers=NULL;  
   char curl_errorstr[CURL_ERROR_SIZE];
   long curl_responsecode;
   URI Uri = Itm->Uri;
   string remotehost = Uri.Host;
   string parsed_uri = Uri;
   
   // S3 Plus hack and s3 URI mod
   
   parsed_uri.erase(0, 3);
   
   string normalized_uri = QuoteString(parsed_uri, "~");
   size_t f_plus = normalized_uri.find("+");
   size_t f_rep;
   while (f_plus != string::npos){
      f_rep = f_plus;
      normalized_uri.replace(f_rep, 1, "%2b");
      f_plus = normalized_uri.find("+", f_plus + 1);
   }

   // Same with path
   string normalized_path = QuoteString(Uri.Path, "~");
   size_t fp_plus = normalized_path.find("+");
   size_t fp_rep;
   while (fp_plus != string::npos){
      fp_rep = fp_plus;
      normalized_path.replace(fp_rep, 1, "%2b");
      fp_plus = normalized_path.find("+", f_plus + 1);
   }

   // TODO:
   //       - http::Pipeline-Depth
   //       - error checking/reporting
   //       - more debug options? (CURLOPT_DEBUGFUNCTION?)

   curl_easy_reset(curl);
   SetupProxy();

   maybe_add_auth (Uri, _config->FindFile("Dir::Etc::netrc"));

   // callbacks
   curl_easy_setopt(curl, CURLOPT_URL, static_cast<string>(normalized_uri).c_str());
   curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_data);
   curl_easy_setopt(curl, CURLOPT_WRITEDATA, this);
   curl_easy_setopt(curl, CURLOPT_PROGRESSFUNCTION, progress_callback);
   curl_easy_setopt(curl, CURLOPT_PROGRESSDATA, this);
   curl_easy_setopt(curl, CURLOPT_NOPROGRESS, false);
   curl_easy_setopt(curl, CURLOPT_FAILONERROR, true);
   curl_easy_setopt(curl, CURLOPT_FILETIME, true);
   // curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);

   // SSL parameters are set by default to the common (non mirror-specific) value
   // if available (or a default one) and gets overload by mirror-specific ones.

   // File containing the list of trusted CA.
   string cainfo = _config->Find("Acquire::https::CaInfo","");
   string knob = "Acquire::https::"+remotehost+"::CaInfo";
   cainfo = _config->Find(knob.c_str(),cainfo.c_str());
   if(cainfo.empty() == false)
      curl_easy_setopt(curl, CURLOPT_CAINFO,cainfo.c_str());

   // Check server certificate against previous CA list ...
   bool peer_verify = _config->FindB("Acquire::https::Verify-Peer",true);
   knob = "Acquire::https::" + remotehost + "::Verify-Peer";
   peer_verify = _config->FindB(knob.c_str(), peer_verify);
   curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, peer_verify);

   // ... and hostname against cert CN or subjectAltName
   bool verify = _config->FindB("Acquire::https::Verify-Host",true);
   knob = "Acquire::https::"+remotehost+"::Verify-Host";
   verify = _config->FindB(knob.c_str(),verify);
   int const default_verify = (verify == true) ? 2 : 0;
   curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, default_verify);

   // Also enforce issuer of server certificate using its cert
   string issuercert = _config->Find("Acquire::https::IssuerCert","");
   knob = "Acquire::https::"+remotehost+"::IssuerCert";
   issuercert = _config->Find(knob.c_str(),issuercert.c_str());
   if(issuercert.empty() == false)
      curl_easy_setopt(curl, CURLOPT_ISSUERCERT,issuercert.c_str());

   // For client authentication, certificate file ...
   string pem = _config->Find("Acquire::https::SslCert","");
   knob = "Acquire::https::"+remotehost+"::SslCert";
   pem = _config->Find(knob.c_str(),pem.c_str());
   if(pem.empty() == false)
      curl_easy_setopt(curl, CURLOPT_SSLCERT, pem.c_str());

   // ... and associated key.
   string key = _config->Find("Acquire::https::SslKey","");
   knob = "Acquire::https::"+remotehost+"::SslKey";
   key = _config->Find(knob.c_str(),key.c_str());
   if(key.empty() == false)
      curl_easy_setopt(curl, CURLOPT_SSLKEY, key.c_str());

   // Allow forcing SSL version to SSLv3 or TLSv1 (SSLv2 is not
   // supported by GnuTLS).
   long final_version = CURL_SSLVERSION_DEFAULT;
   string sslversion = _config->Find("Acquire::https::SslForceVersion","");
   knob = "Acquire::https::"+remotehost+"::SslForceVersion";
   sslversion = _config->Find(knob.c_str(),sslversion.c_str());
   if(sslversion == "TLSv1")
     final_version = CURL_SSLVERSION_TLSv1;
   else if(sslversion == "SSLv3")
     final_version = CURL_SSLVERSION_SSLv3;
   curl_easy_setopt(curl, CURLOPT_SSLVERSION, final_version);

   // CRL file
   string crlfile = _config->Find("Acquire::https::CrlFile","");
   knob = "Acquire::https::"+remotehost+"::CrlFile";
   crlfile = _config->Find(knob.c_str(),crlfile.c_str());
   if(crlfile.empty() == false)
      curl_easy_setopt(curl, CURLOPT_CRLFILE, crlfile.c_str());

   // cache-control
   if(_config->FindB("Acquire::https::No-Cache",
  _config->FindB("Acquire::http::No-Cache",false)) == false)
   {
      // cache enabled
      if (_config->FindB("Acquire::https::No-Store",
    _config->FindB("Acquire::http::No-Store",false)) == true)
   headers = curl_slist_append(headers,"Cache-Control: no-store");
      stringstream ss;
      ioprintf(ss, "Cache-Control: max-age=%u", _config->FindI("Acquire::https::Max-Age",
    _config->FindI("Acquire::http::Max-Age",0)));
      headers = curl_slist_append(headers, ss.str().c_str());
   } else {
      // cache disabled by user
      headers = curl_slist_append(headers, "Cache-Control: no-cache");
      headers = curl_slist_append(headers, "Pragma: no-cache");
   }
   curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

   // speed limit
   int const dlLimit = _config->FindI("Acquire::https::Dl-Limit",
    _config->FindI("Acquire::http::Dl-Limit",0))*1024;
   if (dlLimit > 0)
      curl_easy_setopt(curl, CURLOPT_MAX_RECV_SPEED_LARGE, dlLimit);

   // set header
   curl_easy_setopt(curl, CURLOPT_USERAGENT,
  _config->Find("Acquire::https::User-Agent",
    _config->Find("Acquire::http::User-Agent",
      "Debian S3-CURL/1.0 (" VERSION ")").c_str()).c_str());

   // set timeout
   int const timeout = _config->FindI("Acquire::https::Timeout",
    _config->FindI("Acquire::http::Timeout",120));
   curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, timeout);
   //set really low lowspeed timeout (see #497983)
   curl_easy_setopt(curl, CURLOPT_LOW_SPEED_LIMIT, DL_MIN_SPEED);
   curl_easy_setopt(curl, CURLOPT_LOW_SPEED_TIME, timeout);

   // set redirect options and default to 10 redirects
   bool const AllowRedirect = _config->FindB("Acquire::https::AllowRedirect",
  _config->FindB("Acquire::http::AllowRedirect",true));
   curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, AllowRedirect);
   curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 10);

   // debug
   if(_config->FindB("Debug::Acquire::https", false))
      curl_easy_setopt(curl, CURLOPT_VERBOSE, true);

   // error handling
   curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, curl_errorstr);

   // If we ask for uncompressed files servers might respond with content-
   // negotation which lets us end up with compressed files we do not support,
   // see 657029, 657560 and co, so if we have no extension on the request
   // ask for text only. As a sidenote: If there is nothing to negotate servers
   // seem to be nice and ignore it.
   if (_config->FindB("Acquire::https::SendAccept", _config->FindB("Acquire::http::SendAccept", true)) == true)
   {
      size_t const filepos = Itm->Uri.find_last_of('/');
      string const file = Itm->Uri.substr(filepos + 1);
      if (flExtension(file) == file)
   headers = curl_slist_append(headers, "Accept: text/*");
   }

   // if we have the file send an if-range query with a range header
   if (stat(Itm->DestFile.c_str(),&SBuf) >= 0 && SBuf.st_size > 0)
   {
      char Buf[1000];
      sprintf(Buf, "Range: bytes=%li-", (long) SBuf.st_size - 1);
      headers = curl_slist_append(headers, Buf);
      sprintf(Buf, "If-Range: %s", TimeRFC1123(SBuf.st_mtime).c_str());
      headers = curl_slist_append(headers, Buf);
   }
   else if(Itm->LastModified > 0)
   {
      curl_easy_setopt(curl, CURLOPT_TIMECONDITION, CURL_TIMECOND_IFMODSINCE);
      curl_easy_setopt(curl, CURLOPT_TIMEVALUE, Itm->LastModified);
   }
   
   // S3 fun
   time_t rawtime = 0;
   struct tm * timeinfo = NULL;
   char buffer [80] = { 0 };
   char* wday = NULL;

   time( &rawtime);
   timeinfo = gmtime( &rawtime);

   // strftime does not seem to honour set_locale(LC_ALL, "") or
   // set_locale(LC_TIME, ""). So convert day of week by hand.
   switch (timeinfo->tm_wday) {
   case 0: wday = (char*)"Sun"; break;
   case 1: wday = (char*)"Mon"; break;
   case 2: wday = (char*)"Tue"; break;
   case 3: wday = (char*)"Wed"; break;
   case 4: wday = (char*)"Thu"; break;
   case 5: wday = (char*)"Fri"; break;
   case 6: wday = (char*)"Sat"; break;
   }

   strcat(buffer, wday);
   strftime(buffer+3, 80, ", %d %b %Y %T %z", timeinfo);
   string dateString((const char*)buffer);
   
   headers = curl_slist_append(headers, ("Date: " + dateString).c_str());

   string extractedPassword;
   if(getenv("AWS_SECRET_ACCESS_KEY") != NULL) {
     extractedPassword = getenv("AWS_SECRET_ACCESS_KEY");
   } else if(!Uri.Password.empty()) {
     if(Uri.Password.at(0) == '['){
       extractedPassword = Uri.Password.substr(1,Uri.Password.size()-2);
     }else{
       extractedPassword = Uri.Password;
     }
   }

   char headertext[SLEN], signature[SLEN];
   sprintf(headertext,"GET\n\n\n%s\n%s", dateString.c_str(), normalized_path.c_str());
   doEncrypt(headertext, signature, extractedPassword.c_str());

   string signatureString(signature);
   string user;
   if (getenv("AWS_ACCESS_KEY_ID") != NULL) {
     user = getenv("AWS_ACCESS_KEY_ID");
   } else if(!Uri.User.empty()) {
     user = Uri.User;
   }

   if(!user.empty() || !extractedPassword.empty())
     headers = curl_slist_append(headers, ("Authorization: AWS " + user + ":" + signatureString).c_str());
   // end S3 fun

   // go for it - if the file exists, append on it
   File = new FileFd(Itm->DestFile, FileFd::WriteAny);
   if (File->Size() > 0)
      File->Seek(File->Size() - 1);

   // keep apt updated
   Res.Filename = Itm->DestFile;

   // get it!
   CURLcode success = curl_easy_perform(curl);
   curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &curl_responsecode);

   long curl_servdate;
   curl_easy_getinfo(curl, CURLINFO_FILETIME, &curl_servdate);

   File->Close();

   // cleanup
   if(success != 0) 
   {
      _error->Error("%s", curl_errorstr);
      // unlink, no need keep 401/404 page content in partial/
      unlink(File->Name().c_str());
      Fail();
      return true;
   }

   // Timestamp
   struct utimbuf UBuf;
   if (curl_servdate != -1) {
       UBuf.actime = curl_servdate;
       UBuf.modtime = curl_servdate;
       utime(File->Name().c_str(),&UBuf);
   }

   // check the downloaded result
   struct stat Buf;
   if (stat(File->Name().c_str(),&Buf) == 0)
   {
      Res.Filename = File->Name();
      Res.LastModified = Buf.st_mtime;
      Res.IMSHit = false;
      if (curl_responsecode == 304)
      {
   unlink(File->Name().c_str());
   Res.IMSHit = true;
   Res.LastModified = Itm->LastModified;
   Res.Size = 0;
   URIDone(Res);
   return true;
      }
      Res.Size = Buf.st_size;
   }

   // take hashes
   Hashes Hash;
   FileFd Fd(Res.Filename, FileFd::ReadOnly);
   Hash.AddFD(Fd);
   Res.TakeHashes(Hash);

   // keep apt updated
   URIDone(Res);

   // cleanup
   Res.Size = 0;
   delete File;
   curl_slist_free_all(headers);

   return true;
};

int main()
{
   setlocale(LC_ALL, "");

   HttpsMethod Mth;
   curl_global_init(CURL_GLOBAL_SSL) ;

   return Mth.Run();
}


src/s3+http_main.cc
/*  Copyright (C) 2008,2012 Kyle Shank
    This file is part of apt-s3.

    apt-s3 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    apt-s3 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with apt-s3.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <apt-pkg/fileutl.h>
#include <apt-pkg/acquire-method.h>

#include "connect.h"
#include "rfc2553emu.h"
#include "s3+http.h"
#include <iostream>
#include <fstream>
#include <unistd.h>

using namespace std;

int main()
{
   setlocale(LC_ALL, "");

   HttpMethod Mth;
   return Mth.Loop();
}


src/s3+http.h
/*  Copyright (C) 2008,2012 Kyle Shank
    This file is part of apt-s3.

    apt-s3 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    apt-s3 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with apt-s3.  If not, see <http://www.gnu.org/licenses/>.
*/
// -*- mode: cpp; mode: fold -*-
// Description								/*{{{*/// $Id: http.h,v 1.12 2002/04/18 05:09:38 jgg Exp $
// $Id: http.h,v 1.12 2002/04/18 05:09:38 jgg Exp $
/* ######################################################################

   HTTP Acquire Method - This is the HTTP aquire method for APT.

   ##################################################################### */
									/*}}}*/

#ifndef APT_HTTP_H
#define APT_HTTP_H

#include <apt-pkg/hashes.h>

#define MAXLEN 360



using std::cout;
using std::endl;

class HttpMethod;

class CircleBuf
{
   unsigned char *Buf;
   unsigned long Size;
   unsigned long InP;
   unsigned long OutP;
   string OutQueue;
   unsigned long StrPos;
   unsigned long MaxGet;
   struct timeval Start;
   
   static unsigned long BwReadLimit;
   static unsigned long BwTickReadData;
   static struct timeval BwReadTick;
   static const unsigned int BW_HZ;

   unsigned long LeftRead()
   {
      unsigned long Sz = Size - (InP - OutP);
      if (Sz > Size - (InP%Size))
	 Sz = Size - (InP%Size);
      return Sz;
   }
   unsigned long LeftWrite()
   {
      unsigned long Sz = InP - OutP;
      if (InP > MaxGet)
	 Sz = MaxGet - OutP;
      if (Sz > Size - (OutP%Size))
	 Sz = Size - (OutP%Size);
      return Sz;
   }
   void FillOut();
   
   public:
   
   Hashes *Hash;
   
   // Read data in
   bool Read(int Fd);
   bool Read(string Data);
   
   // Write data out
   bool Write(int Fd);
   bool WriteTillEl(string &Data,bool Single = false);
   
   // Control the write limit
   void Limit(long Max) {if (Max == -1) MaxGet = 0-1; else MaxGet = OutP + Max;}   
   bool IsLimit() {return MaxGet == OutP;};
   void Print() {cout << MaxGet << ',' << OutP << endl;};

   // Test for free space in the buffer
   bool ReadSpace() {return Size - (InP - OutP) > 0;};
   bool WriteSpace() {return InP - OutP > 0;};

   // Dump everything
   void Reset();
   void Stats();

   CircleBuf(unsigned long Size);
   ~CircleBuf() {delete [] Buf; delete Hash;};
};

struct ServerState
{
   // This is the last parsed Header Line
   unsigned int Major;
   unsigned int Minor;
   unsigned int Result;
   char Code[MAXLEN];
   
   // These are some statistics from the last parsed header lines
   unsigned long Size;
   signed long StartPos;
   time_t Date;
   bool HaveContent;
   enum {Chunked,Stream,Closes} Encoding;
   enum {Header, Data} State;
   bool Persistent;
   
   // This is a Persistent attribute of the server itself.
   bool Pipeline;
   
   HttpMethod *Owner;
   
   // This is the connection itself. Output is data FROM the server
   CircleBuf In;
   CircleBuf Out;
   int ServerFd;
   URI ServerName;
  
   bool HeaderLine(string Line);
   bool Comp(URI Other) {return Other.Host == ServerName.Host && Other.Port == ServerName.Port;};
   void Reset() {Major = 0; Minor = 0; Result = 0; Size = 0; StartPos = 0;
                 Encoding = Closes; time(&Date); ServerFd = -1; 
                 Pipeline = true;};
   int RunHeaders();
   bool RunData();
   
   bool Open();
   bool Close();
   
   ServerState(URI Srv,HttpMethod *Owner);
   ~ServerState() {Close();};
};

class HttpMethod : public pkgAcqMethod
{
   void SendReq(FetchItem *Itm,CircleBuf &Out);
   bool Go(bool ToFile,ServerState *Srv);
   bool Flush(ServerState *Srv);
   bool ServerDie(ServerState *Srv);
   int DealWithHeaders(FetchResult &Res,ServerState *Srv);

   virtual bool Configuration(string Message);
   
   // In the event of a fatal signal this file will be closed and timestamped.
   static string FailFile;
   static int FailFd;
   static time_t FailTime;
   static void SigTerm(int);

   protected:
   virtual bool Fetch(FetchItem *);
   
   public:
   friend class ServerState;

   FileFd *File;
   ServerState *Server;
   
   int Loop();
   
   HttpMethod() : pkgAcqMethod("1.2",Pipeline | SendConfig) 
   {
      File = 0;
      Server = 0;
   };
};

#endif
